### PR TITLE
Use xstate interpreter

### PR DESCRIPTION
**Pull request by LevelbossMike**
Xstate now comes with [interpreter functionality](https://xstate.js.org/docs/guides/interpretation.html#interpreter) out of the box. Making use of xstate's interpreter makes sure users of `ember-statecharts` can use all of xstate's functionality, e.g. [history-state](https://xstate.js.org/docs/guides/history.html) wasn't supported before nor were new features like [activities](https://xstate.js.org/docs/guides/activities.html) or [delays](https://xstate.js.org/docs/guides/delays.html#delayed-transitions). By relying on xstate's native interpreter we make sure we get new features xstate implements for free in the future.

Questions regarding this PR:

- ~~we are not using Ember's runloop for the [interpreter's clock](https://xstate.js.org/docs/guides/delays.html#interpretation)-functionality. We most likely want to use the runloop though as it will make certain things easier to test.~~
- this PR unfortunately changes the statechart api for `ember-statecharts` - we are now not binding `this` of actions to the `Ember.Object` that is implementing the statechart. This will create the need for users to update their statechart configurations if they rely on this behavior. The benefit is that `actions` and `guards` now behave the same way which is easier to get for new users.
- ~~this PR keeps the custom `send`-interface on statecharts - this format is specific to `ember-statecharts` and diverges from the way developers can send event's with xstate. xstate allows their users to send events by [sending an `event-object`](https://xstate.js.org/docs/guides/events.html#sending-events). I feel like the way `ember-statecharts` lets users send event-data - i.e. `statechart.send('EVENTNAME', { woot: "fubar" })` and then making the event-data accessible in actions via the data-property is more conventional - i.e. `wat(ctx, { data: { woot } }) => console.log(woot)`. If we feel this is unecessary and we want to mirror xstate's `send` that's fine for me as well. People will need to update their statechart configurations anyway so we could also force them to follow xstate as closely as possible from now on.~~
- ~~I added prettier to the addon to make sure code stays consistent. This obviously should have gone into a separate PR - I can pull changes out and do a follow up PR if need be~~
- `statechart#send` now aligns with what xstate does (and will do from 4.5 on) - i.e. you can either pass a string, an event-object (`{ type: <event-name>, ...data }`) or pass two params `eventName`, and additional event-data.

### CODE REVIEW

**Comment by LevelbossMike** [comment]
Decided to remove prettier changes as it makes hard to follow what's going on in this pr

**Comment by davidkpiano** [comment]
> I feel like the way ember-statecharts lets users send event-data - i.e. statechart.send('EVENTNAME', { woot: "fubar" }) and then making the event-data accessible in actions via the data-property is more conventional

You will also be able to do this in the upcoming XState version 4.5.

**Comment by LevelbossMike** [comment]
Updated as discussed in our call @pangratz:

Send was changed in the following way:

* add support for xstate event-objects when only passing one param to `send`
* issue warning when type property would override event-name that was passed to `send`
* use passed event-name instead of `type` from data passed to `send`

File: addon/utils/statechart.js [hunk 1]
```diff
@@ -1,149 +1,27 @@
...
-            on: {
-              send: {
-                target: 'initialized',
-                actions: ['executeSend'],
-              },
-            },
-          },
+  constructor(config, options, initialContext) {
+    const machine = Machine(config, options, initialContext);
+    this.service = interpret(machine, {
```
[the remark this answers is not in the document]
Comment (davidkpiano) [reply addon/utils/statechart.js]: Easier to think about it as a service.

File: addon/utils/statechart.js [hunk 2]
```diff
@@ -1,149 +1,45 @@
...
+      const { type, ...eventData } = data;
+      if (type) {
+        warn(
+          `You passed property \`type\` as part of the data you sent with the event \`${event}\` . This is not supported - \`${event}\` will be used as event name.`,
+          false,
+          { id: 'statecharts.event-object.no-override-type' }
+        );
+      }
+      const eventObject = { ...eventData, type: event };
+      this.service.send(eventObject);
```
Comment (pangratz) [inline addon/utils/statechart.js]: Call `me._sendEventObkject` maybe ☎️ ?
